### PR TITLE
EVG-5477 support pulling docker images from registries

### PR DESCRIPTION
**Pull request by john-m-liu**
the earliest commit is probably the only one worth reviewing

### CODE REVIEW

File: cloud/docker_client.go [hunk 1]
```diff
@@ -149,57 +152,99 @@ func (c *dockerClientImpl) EnsureImageDownloaded(ctx context.Context, h *host.Ho
...
+				"duration":  time.Since(start),
+				"span":      time.Since(start).String(),
+			})
+			err = c.pullImage(ctx, h, url, settings["docker_registry_user"].(string), settings["docker_registry_pw"].(string))
+			grip.Info(message.Fields{
+				"operation": "EnsureImageDownloaded",
+				"details":   "done pulling image",
+				"duration":  time.Since(start),
+				"span":      time.Since(start).String(),
+			})
```
Comment (tychoish) [inline cloud/docker_client.go]: maybe it makes sense to only log the end? or refactor out the start? 

I know I'm guilty of a lot of duration+span but I think putting duration.Seconds() is reasonable.

File: cloud/docker_client.go [hunk 2]
```diff
@@ -149,57 +152,99 @@ func (c *dockerClientImpl) EnsureImageDownloaded(ctx context.Context, h *host.Ho
...
+			return imageName, errors.Wrap(err, "error pulling image")
 		}
+		return imageName, errors.Errorf("unrecognized image build method: %s", (*h.Distro.ProviderSettings)["build_type"])
```
Comment (tychoish) [inline cloud/docker_client.go]: I feel like getting (*h.DistroProviderSettings)["build_id"] to be an error but not panic seems slim. maybe something narrower.

File: cloud/docker_client.go (hunk 1 above)
Comment (john-m-liu) [reply cloud/docker_client.go]: changed

File: cloud/docker_client.go (hunk 2 above)
Comment (john-m-liu) [reply cloud/docker_client.go]: oops, fixed

File: self-tests.yml [hunk 3]
```diff
@@ -201,6 +214,32 @@ functions:
...
+        timeout_seconds: 600
+        num_hosts: 1
+        path: spawned_hosts.json
+    - command: shell.exec
+      params:
+        script: |
+          set -o xtrace
+          set -o errexit
+          hostname=$(tr -d '"[]{}' < spawned_hosts.json | cut -d , -f 1 | awk -F : '{print $2}' | head -3 | xargs)
+          printf "docker_host: %s" "$hostname" > expansions.yml
```
[the remark this answers is not in the document]
Comment (john-m-liu) [reply self-tests.yml]: done

**Pull request merged**